core/backend/src/test/annotations/TextAnnotation.test.ts

```diff
@@ -551,6 +551,242 @@ describe("layoutTextBlock", () => {
+        const resultRun: RunLayoutResult = resultLine.runs[j];
+        const originalRun: RunLayout = originalLine.runs[j];
+
+        // Source run index matches
+        expect(resultRun.sourceRunIndex).to.equal(textBlock.paragraphs[resultLine.sourceParagraphIndex].runs.indexOf(originalRun.source));
+        // FontId matches
+        expect(resultRun.fontId).to.equal(originalRun.fontId);
+        // Offsets match
+        expect(resultRun.characterOffset).to.equal(originalRun.charOffset);
+        expect(resultRun.characterCount).to.equal(originalRun.numChars);
+        expect(resultRun.offsetFromLine).to.deep.equal(originalRun.offsetFromLine);
+        // Range matches
+        expect(resultRun.range).to.deep.equal(originalRun.range.toJSON());
+        // Text style matches
+        expect(resultRun.textStyle).to.deep.equal(originalRun.style.toJSON());
+        // Optional values match existence and values
+        if (resultRun.justificationRange) {
+          expect(originalRun.justificationRange);
+        }
+        if (originalRun.justificationRange) {
+          expect(resultRun.justificationRange);
+        }
+        if (resultRun.justificationRange && originalRun.justificationRange) {
+          expect(resultRun.justificationRange).to.deep.equal(originalRun.justificationRange.toJSON());
+        }
+        if (resultRun.numeratorRange) {
+          expect(originalRun.numeratorRange);
+        }
+        if (originalRun.numeratorRange) {
+          expect(resultRun.numeratorRange);
+        }
+        if (resultRun.numeratorRange && originalRun.numeratorRange) {
+          expect(resultRun.numeratorRange).to.deep.equal(originalRun.numeratorRange.toJSON());
+        }
+        if (resultRun.denominatorRange) {
+          expect(originalRun.denominatorRange);
+        }
+        if (originalRun.denominatorRange) {
+          expect(resultRun.denominatorRange);
+        }
+        if (resultRun.denominatorRange && originalRun.denominatorRange) {
+          expect(resultRun.denominatorRange).to.deep.equal(originalRun.denominatorRange.toJSON());
+        }
+        // Check that the result string matches what we expect
+        const inputRun = textBlock.paragraphs[resultLine.sourceParagraphIndex].runs[resultRun.sourceRunIndex].clone();
+        if (inputRun.type === "text") {
+          const resultText = inputRun.content.substring(resultRun.characterOffset, resultRun.characterOffset + resultRun.characterCount);
+          const originalText = inputRun.content.substring(originalRun.charOffset, originalRun.charOffset + originalRun.numChars);
+          expect(resultText).to.equal(originalText);
+        }
+      }
+    }
+  });
+
+  describe("grapheme offsets", () => {
+    it("should return an empty array if source type is not text", function () {
+      const textBlock = TextBlock.create({ styleName: "" });
+      const fractionRun = FractionRun.create({ numerator: "1", denominator: "2", styleName: "fraction" });
```

Comment:
What does/will your text editing tool do when user tries to move caret over a fraction? Is it possible to edit or delete a fraction?

Reply:
Right now, the caret API treats a fraction as an atomic unit. It uses the range of the FractionRun to determine the cursor's offset. 

Fractions are not yet implemented with the text editing tool. With what I have right now, it would not be possible to edit a fraction. It would be possible to delete a fraction (this would be the same as removing the entire FractionRun). 

I think it would be very nice to be able to edit fractions in-place, but I have not started thinking about how to implement that.

Reply:
I mainly ask because it may have implications for your new function and interfaces. They are beta, so they can evolve as you iterate, but it may be worth taking some time to consider in what direction they might need to evolve to support fractions, so you don't have to radically revise them later.

Reply:
Spending only the last 30 minutes thinking about it, I put together some pseudo code that should allow the function and interface to remain mostly the same for fractions. I would need to add one more property to the interface to know if the function should get grapheme range's for the numerator or denominator (meaning for each fraction the function would need to be called twice, once for the numerator and once for the denominator). I'm not sure I love that, but it should be possible.

However, there are a lot of big unknowns with text (mainly around international language support), and I am viewing this function and interface as the first steps towards validating those unknowns. I think it is too early to conclude if this approach is “good enough” or not, so it will probably change regardless as we iterate